src/util/threadpool.rs

```diff
@@ -569,4 +665,112 @@ mod test {
         assert_eq!(task.gid, group2);
         queue.on_task_started(&group2);
     }
+
+
+    #[test]
+    fn test_small_group_first_queue() {
+        let concurrency_limit = 2;
+        let small_group_tasks_limit = 2;
+        let mut queue = SmallGroupFirstQueue::new(concurrency_limit, small_group_tasks_limit);
+        let mut id = 1;
+        let mut pre_id = 0;
+
+        // high_priority_queue and low_priority_queue is both empty.
+        assert!(queue.pop().is_none());
+
+        // Push 2 tasks of `group1` into queue
+        let group1 = 1001;
+        for _ in 0..small_group_tasks_limit {
+            let mut task = Task::new(group1, move || {});
+            task.id = id;
+            id += 1;
+            queue.push(task);
+        }
+
+        // high:[g11,g12]; low:[]
+        assert!(queue.low_priority_queue.is_empty());
+
+        // low_priority_queue is empty
+        let task = queue.pop().unwrap();
+        assert!(task.id > pre_id);
+        pre_id = task.id;
+        queue.on_task_started(&task.gid);
+
+        // high:[g12]; low:[];running:g11
+        for _ in 0..small_group_tasks_limit {
+            let mut task = Task::new(group1, move || {});
+            task.id = id;
+            id += 1;
+            queue.push(task);
+        }
+
+        // since g11(front in high_queue) comes before g13(front in high_queue),
+        // g11 would run first. high:[g12]; low:[g13,g14];running:[g11]
+        assert_eq!(queue.low_priority_queue.len(), small_group_tasks_limit);
+        assert_eq!(queue.high_priority_queue.len(), small_group_tasks_limit - 1);
+
+        let group2 = 1002;
+        for _ in 0..small_group_tasks_limit {
+            let mut task = Task::new(group2, move || {});
+            task.id = id;
+            id += 1;
+            queue.push(task);
+        }
+        // high:[g12,g21,g22], low:[g13,g14],running:[g11]
+        let task = queue.pop().unwrap();
+        assert!(task.id > pre_id);
+        queue.on_task_started(&task.gid);
+        // since g12(front in high_queue) comes before g13(front in low_queue),
+        // g12 would run first.
+        assert_eq!(queue.low_priority_queue.len(), small_group_tasks_limit);
+
+        // high:[g21,g22],low:[g13,g14],running:[g11,g12]
+        queue.on_task_finished(&task.gid);
+        // high:[g21,g22],low:[g13,g14],running:[g11]
+        let task = queue.pop().unwrap();
+        queue.on_task_started(&task.gid);
+        // since g13(font in low_queue) comes before g21(front in high_queue),
+        // and group g1's running number < group_concurrency_on_busy(2)
+        // g13 would run first.
+        assert_eq!(task.gid, group1);
+
+        // high:[g21,g22],low:[g14], running:[g11,g13]
+        let task = queue.pop().unwrap();
+        queue.on_task_started(&task.gid);
+        // since group g1's running number >= 2(group_concurrency_on_busy),
```

Comment:
It should be consistent to write comment in either way `2(group_concurrency_on_busy)` in this line or `group_concurrency_on_busy(2)` in line 733.